Add support for multiple inputs and outputs, GPU and reshape

CoderHam opened the pull request:
- [x] Add GPU Support
- [x] Support multiple I/Os
- [x] Support reshape
- [x] Use CopyBuffer() util
- [x] Add support for params execute_delay_ms and instance_wise_delay_multiplier

PS: Previously supported only 1 input and 1 output on CPU

CoderHam committed:
Extend support for multiple inputs and outputs
- Currently supported only 1 input and 1 output

GuanLuo commented:
Why do we need this change?

CoderHam commented:
@GuanLuo we need this to replace the legacy custom backend models used in the ensembles (nop_*) with the new identitiy backend models since we are removing the legacy custom backend. 
The nop models are part of the ensembles used in L0_infer (among other places)

GuanLuo commented on src/identity.cc:
```diff
@@ -210,40 +210,37 @@ ModelState::ValidateModelConfig()
+      inputs.ArraySize() == outputs.ArraySize(), TRITONSERVER_ERROR_INVALID_ARG,
+      std::string("model configuration must have equal input/output pairs"));
+
+  for (size_t io_index = 0; io_index < inputs.ArraySize(); io_index++) {
```
Identity in multiple I/O case maps the input to output based on the naming (i.e. "INPUT0" to "OUTPUT0"). You shouldn't assume the matching input and output are at the same index, you may have a config like below while we actually want to validate the spec between "INPUT0" and "OUTPUT0"
```
input [
{ name: "INPUT2" ...},
{ name: "INPUT0" ...}, 
{ name: "INPUT1" ...}
]
output [
{ name: "OUTPUT0" ...},
{ name: "OUTPUT1" ...}, 
{ name: "OUTPUT2" ...}
]
```

GuanLuo replied:
Have your new changes addressed this [comment](https://github.com/triton-inference-server/identity_backend/pull/9#discussion_r608362003)? What will your current change do to validate the config above?

CoderHam replied:
Fixed by latest commit

GuanLuo commented on src/identity.cc:
```diff
@@ -680,175 +673,178 @@ TRITONBACKEND_ModelInstanceExecute(
-          TRITONBACKEND_ResponseOutput(
-              response, &output, requested_output_name, input_datatype,
-              input_shape, input_dims_count));
+          TRITONBACKEND_RequestInput(request, input_name, &input));
```
You can get the input directly by using `TRITONBACKEND_RequestInputByIndex`, that way you need to get the input name via `TRITONBACKEND_InputProperties` at line 718, but you save one API call

GuanLuo commented on src/identity.cc:
```diff
@@ -680,175 +673,178 @@ TRITONBACKEND_ModelInstanceExecute(
+        GUARDED_RESPOND_IF_ERROR(
+            responses, r,
+            TRITONBACKEND_RequestOutputName(
+                request, io_index, &requested_output_name));
```
Same idea as what I mentioned in validation

GuanLuo commented:
Can you also update the places where this backend will be used so your changes can be tested?

CoderHam commented:
> Can you also update the places where this backend will be used so your changes can be tested?

Sure

CoderHam committed:
review edits

CoderHam committed:
Add GPU support

CoderHam committed:
Add support for reshape

deadeyegoodwin commented on src/identity.cc:
```diff
@@ -26,6 +26,7 @@
 
 #include <memory>
 #include <thread>
+
```
Don't put a blank line between system and local includes

deadeyegoodwin commented on src/identity.cc:
```diff
@@ -26,6 +26,7 @@
+
 #include "triton/backend/backend_common.h"
 
 namespace triton { namespace backend { namespace identity {
```
Need to change description below...

CoderHam replied:
Fixed

deadeyegoodwin commented on src/identity.cc:
```diff
@@ -680,175 +694,288 @@ TRITONBACKEND_ModelInstanceExecute(
+                .c_str());
+      }
+
+      // For statistics we need to collect the total batch size of all
```
This hsouldn't be inside the output loop, should it? You are multi-counting batch size for each output.

deadeyegoodwin commented on src/identity.cc:
```diff
@@ -680,175 +694,288 @@ TRITONBACKEND_ModelInstanceExecute(
-    const char* requested_output_name = nullptr;
-    if (requested_output_count > 0) {
+    // Collect all requested outputs
+    std::vector<const char*> requested_output_names;
```
Put these in a set to make it easier to determine if one is specified

deadeyegoodwin commented on src/identity.cc:
```diff
@@ -680,175 +694,288 @@ TRITONBACKEND_ModelInstanceExecute(
+      const char* requested_output_name = nullptr;
+      for (uint32_t io_index = 0; io_index < requested_output_count;
+           io_index++) {
+        if (output_name.compare(requested_output_names[io_index]) == 0) {
```
if set.contains(output_name) then can just continue? No need for is_requested flag?

GuanLuo replied:
I think you should handle this backward, instead of iterating "input -> check if corresponding output is requested -> copy to output", you should just iterate "requested output -> look up corresponding input -> copy to output".

Then you don't need to maintain a set of requested output

CoderHam replied:
Yes, thats a good idea. Did that

deadeyegoodwin commented on src/identity.cc:
```diff
@@ -680,175 +694,288 @@ TRITONBACKEND_ModelInstanceExecute(
+
+      // We only need to produce an output if it was requested.
+      if (is_requested) {
+        // This backend simply copies the input tensor to the output
```
We should be able to use the BackendResponder utilities for this. That will also allow you to not have to manage the cuda memcpy, etc. below.

GuanLuo replied:
I think `CopyBuffer()` in backend_common.h can be used. The responder is operating on the whole batch, and it expects caller to provide it a contiguous output buffer from the model to be partitioned into individual responses, which seems to be an overkill as we can handle each request individually.

CoderHam replied:
Fixed to use CopyBuffer()

CoderHam committed:
improvements based on review comments

CoderHam committed:
Additional fixes
- needed to use string instead of char* because set.find does not work as expected

CoderHam committed:
validate input and output index convention

CoderHam committed:
Use CopyBuffer from backend_common

deadeyegoodwin commented:
@CoderHam  This change should be able to stand on its own. So can you run a CI with just this change and not all the other legacy custom backend removal changes. Then we can merge this.

CoderHam committed:
use map to compare name, shape and dtype correctly

CoderHam committed:
add compute timestamp flags

CoderHam committed:
Add support for instance_wise_delay_multiplier and execute_delay_ms

CoderHam committed:
Update readme for GPU enabled build

CoderHam committed:
Fix extraction of params
- execute_delay_ms and instance_wise_delay_multiplier

GuanLuo commented on src/identity.cc:
```diff
@@ -275,17 +369,24 @@ class ModelInstanceState {
+  // Returns the stream associated with this instance that can be used
+  // for GPU<->CPU memory transfers. Returns nullptr if GPU support is
+  // disabled or if this instance is not executing on a GPU.
+  cudaStream_t CudaStream() { return stream_; }
```
You can get this for free if you inherit from `BackendModelInstance` class

GuanLuo replied:
And the create / destroy below is also handled.

CoderHam replied:
We have extra components and restrictions in BackendModelInstance and BackendModel (for ModelState) that we do not need to impose here.

GuanLuo commented on src/identity.cc:
```diff
@@ -592,6 +718,17 @@ TRITONBACKEND_ModelInstanceExecute(
+  if (model_state->ExecDelay() > 0) {
+    int multiplier = 1;
+    if (model_state->Multiplier() > 0) {
+      multiplier *= instance_state->DeviceId();
```
The legacy implementation is not setting multiplier based on device id, instead, it is based on the instance id because device id will always be -1 for CPU instance and that can't be used to set different delays for different instances

CoderHam replied:
I used DeviceId since we set device_id to instance_id in the ModelInstanceState constructor

CoderHam replied:
@deadeyegoodwin I think you had done this. Was there a reason for setting instance_id to device_id?

deadeyegoodwin replied:
I didn't change this code from "instance id" to "device id", did I?

deadeyegoodwin replied:
> 
> 
> I used DeviceId since we set device_id to instance_id in the ModelInstanceState constructor

That's not right. Can you point to that code.

CoderHam replied:
I think I misunderstood earlier. 
https://github.com/triton-inference-server/identity_backend/blob/main/src/identity.cc#L314 is the constructor

https://github.com/triton-inference-server/identity_backend/blob/main/src/identity.cc#L308-L310 is the call that initialized device_id_ but although instance_id is the name of the variable it is set using TRITONBACKEND_ModelInstanceDeviceId and is likely the device id and not instance id (correct?)

deadeyegoodwin replied:
Model instances have unique names but they are not assigned unique ids. The backend API lists everything you can know about an instance: https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonbackend.h#L787

What is "Multiplier" used for. It is strange to multiply it by either instance id (if there was such a thing), or device id.

GuanLuo replied:
It is used to make different instances have different execution delay so the later requests are finished early in the test, so that we can check if order is preserved in scheduler even if that happens

CoderHam replied:
This carries over from the legacy custom backend. we have an instance_wise_delay_multiplier which is what Multiplier here refers to https://github.com/triton-inference-server/identity_backend/blob/hemantj-identity/src/identity.cc#L328-L339

The corresponding section to the custom backend is here: https://github.com/triton-inference-server/server/blob/master/src/test/custom/identity/identity.cc#L124-L143
There is logic in the legacy custom backend here: https://github.com/triton-inference-server/server/blob/d71a2d8c70fe0b58ebf187b079d0ccd02b327a52/src/backends/custom/custom_backend.cc#L122-L134 that creates the name using the "instance id"

It is used inside L0_batcher here: https://github.com/triton-inference-server/server/blob/d71a2d8c70fe0b58ebf187b079d0ccd02b327a52/qa/L0_batcher/test.sh#L422

deadeyegoodwin replied:
OK, inside identity model state you will just have to keep an instance count that you can update each time ModelInstanceInitialize is called and then assign that as the unique instance id.

CoderHam replied:
Done

GuanLuo commented on src/identity.cc:
```diff
@@ -786,9 +991,9 @@ TRITONBACKEND_ModelInstanceExecute(
-      // back a buffer in GPU memory we just fail the request.
+      // Step 2. Get the output buffer. We request a buffer in CPU memory but we
+      // have to handle any returned type. If we get back a buffer in GPU memory
+      // we just fail the request.
```
Is this comment still valid? We can handle output on GPU

CoderHam replied:
Fixed

GuanLuo commented on src/identity.cc:
```diff
@@ -826,36 +1046,84 @@ TRITONBACKEND_ModelInstanceExecute(
+              TRITONSERVER_LOG_ERROR,
+              (std::string("request ") + std::to_string(r) +
+               ": failed to get input buffer in CPU memory, error "
+               "response sent")
```
Same as above.

GuanLuo replied:
Unless we actually have existing tests that rely on the assumption that identity can only handle CPU I/O, but in such case, there is not reason to support GPU instance as well?

CoderHam replied:
We do have tests that rely on GPU I/O which is why I had to add support for this

GuanLuo replied:
Add support for what? "Returning error when I/O is on GPU"? Then that means we don't want identity to support GPU instance, right?

CoderHam replied:
To clarify, _this_ meant GPU support. Which was needed for the ensembles in L0_infer (and a few other places that use the nop_ model in ensembles. However we still want the identity backend to be built without GPU support hence the ifdef and the related checks.

GuanLuo replied:
I see, I missed that the macro is `ifndef` instead of `ifdef`. By the way, this will be checked in `CopyBuffer()` so you don't need this if you always use `CopyBuffer()`

CoderHam replied:
Removed

GuanLuo replied:
I meant line 1053 to 1068 could be removed. `CopyBuffer()` will return error if not build with GPU enable and either input / output is with CPU type.

CoderHam replied:
Right, removed these, forgot to push the last commit

GuanLuo commented on src/identity.cc:
```diff
@@ -826,36 +1046,84 @@ TRITONBACKEND_ModelInstanceExecute(
         memcpy(
             reinterpret_cast<char*>(output_buffer) + output_buffer_offset,
             input_buffer, buffer_byte_size);
+#endif  // TRITON_ENABLE_GPU
```
Don't need the `ifdef` and `memcpy`, `CopyBuffer` can handle copy with / without GPU support enabled

CoderHam replied:
A simple memcpy is enough when GPU support is disabled. The CopyBuffer call and checks will add unnecessary slowdown

GuanLuo replied:
You are adding extra complexity here. And is CopyBuffer actually too slow that you have to add the complexity? Checking if it is H2H copy is the first thing it does, so it just one extra function invocation if GPU support is disabled

CoderHam replied:
I did not check for the slowdown but I understand your point. Made these changes

GuanLuo commented on src/identity.cc:
```diff
@@ -826,36 +1046,84 @@ TRITONBACKEND_ModelInstanceExecute(
-      }
+    uint64_t compute_end_ns = 0;
+    SET_TIMESTAMP(compute_end_ns);
+    max_compute_end_ns = std::max(max_compute_end_ns, compute_end_ns);
```
You don't need to compare for the max timestamp because the `compute_end_ns` will always be more recent that `max_compute_end_ns`

CoderHam committed:
review edits

CoderHam committed:
Fix typo

CoderHam committed:
Remove unnecessary memcpy, always use CopyBuffer

CoderHam committed:
use instance id instead of device id

CoderHam committed:
Remove memory checks for no-GPU build
- It is handled by CopyBuffer()

CoderHam committed:
Refactor to reduce repeated code

CoderHam committed:
Fix missing symbol issue

CoderHam committed:
Fix segfault

CoderHam committed:
Multiplier() -> DelayMultiplier()

GuanLuo reviewed (changes requested):
placeholder review so that you can request for review again once the previous comments are addressed

GuanLuo commented on src/identity.cc:
```diff
@@ -1114,12 +957,6 @@ TRITONBACKEND_ModelInstanceExecute(
-    if (cuda_copy) {
-      cudaStreamSynchronize(instance_state->CudaStream());
-    }
-#endif  // TRITON_ENABLE_GPU
```
You still need to synchronize before sending the response when CUDA copy is used, otherwise the PR looks good to me

CoderHam committed:
cudaStreamSynchronize is still needed

CoderHam committed:
Fix timestamp collection

CoderHam committed:
Fix timing when exec delay is added

CoderHam committed:
Fix timestamp collection and instance wise delay multiplication
